libopenage/datastructure/pairing_heap.h

```diff
@@ -583,10 +583,9 @@ class PairingHeap final {
 		this->walk_tree(this->root_node, func);
 	}
 
-protected:
+private:
 	void walk_tree(const element_t &root,
 	               const std::function<void(const element_t &)> &func) const {
-		func(root);
```

Comment:
Ah I didn't thought of that we have top do this for the deletion.

It's probably not a good idea to change the function like this because it effectively reverses the order of walking the tree (`root` will now be the last node that `func` is applied to). I guess the best solution then is to leave this function as-is and not use `iter_all`.

Instead, you can use the code you have written here in `clear()` but replace `func(root)` with `delete root`

Reply:
How about adding a reverse argument to walk tree?

Reply:
Yes, I think that is also a nice option.

I pushed a commit to your branch that additionally makes the reversing a compile-time option, making it faster to execute.